posts/migrating-a-large-postgres-db-from-self-managed-ec2-to-rds.md

```diff
@@ -0,0 +1,115 @@
+
+## Using pg_basebackup
+
+`Pg_basebackup` cannot be used to migrate to AWS’s RDS, but can be used to migrate from one self-managed EC2 instance to a different one. I mention it here simply for thoroughness, as you are likely to come across this method in your research. `pg_basebackup` creates a *physical* backup, at the file system level. This tends to be much faster than the logical backups that I will describe below. The reason this cannot be used to migrate to RDS is that the underlying filesystem is not exposed to the end user when using AWS’s RDS.
+
+It is also worth noting, that on large databases, this method will require a decent length maintenance window, as the database needs to be offline for it to be successful. 
+
+## Using pg_dump and pg_restore
+
+`Pg_dump` and `pg_restore` will create a logical backup of your database. Instead of moving the actual data on the disk, it will use SQL-like statements to completely rebuild the database. There are loads of different options when employing this method. For instance, you can use `pg_dump` to dump your data into .csv files or have the files compressed. Once the data has been dumped onto the hard disk, it can then be loaded into the new database with `pg_restore`.
+
+Another strategy utilizing `pg_dump` involves setting the output to stdin and piping it into SQL copy statements, so it doesn't hit the hard disk at all. This is helpful for really large databases where there may not be enough remaining room on disk, or where time is of the essence and it is better to just send the data over the network into the target database as it is being dumped. This strategy can also be used to migrate tables that may not be compatible with the `.csv` format.
+
+When utilizing this strategy, you will want to make sure there are no schema differences between the source and target database. This strategy can be used to perform the entire migration, if you are allowed a long enough maintenance window. This strategy can also be used in the hybrid approach described below, to migrate *most* of the data, and then use some form of replication to get the last little bit (primarily for migrations that are not allotted a significant length maintenance window).
+
+## Using pg_logical
+
+Logical replication is an amazing tool that can be used to create read replicas of databases and can stream data near real-time from a primary to a secondary database. It leverages the underlying native technology of Postgresql to achieve this. It can also be used to sync databases and perform migrations. There are loads of different options and methodologies that can be used to set this up and utilize it for migrations. The primary benefit is migrating a live database with little to no downtime, resulting in a very short maintenance window for a "cut over".
```

Comment:
```suggestion
Logical replication is an amazing tool that can be used to create read replicas of databases and can stream data near real-time from a primary to a secondary database. It leverages the underlying native technology of Postgresql to achieve this. It can also be used to sync databases and perform migrations. There are different options and methodologies that can be used to set this up and utilize it for migrations. The primary benefit is migrating a live database with little to no downtime, resulting in a very short maintenance window for a "cut over".
```